### PR TITLE
Dependency injection

### PR DESCRIPTION
**Pull request by cl-shaharshavit-rnd**

### CODE REVIEW

File: MergerLogic/DataTypes/Fs.cs [hunk 1]
```diff
@@ -3,17 +3,21 @@
...
 namespace MergerLogic.DataTypes
 {
-    public class FS : Data
+    public class FS : Data<IFileUtils>
     {
         private delegate string TilePathFunction(string path, Tile tile);
 
         private IEnumerator<Tile> tiles;
         private bool done;
         private int completedTiles;
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Fs.cs:12]: _completedTiles

File: MergerLogic/DataTypes/Fs.cs [hunk 2]
```diff
@@ -3,17 +3,21 @@
 
 namespace MergerLogic.DataTypes
 {
-    public class FS : Data
+    public class FS : Data<IFileUtils>
     {
         private delegate string TilePathFunction(string path, Tile tile);
 
         private IEnumerator<Tile> tiles;
         private bool done;
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Fs.cs:11]: _done

File: MergerLogic/DataTypes/Fs.cs [hunk 3]
```diff
@@ -3,17 +3,21 @@
 
 namespace MergerLogic.DataTypes
 {
-    public class FS : Data
+    public class FS : Data<IFileUtils>
     {
         private delegate string TilePathFunction(string path, Tile tile);
 
         private IEnumerator<Tile> tiles;
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/Fs.cs:10]: _tiles

File: MergerLogic/DataTypes/TMS.cs [hunk 4]
```diff
@@ -1,12 +1,14 @@
 ﻿using MergerLogic.Batching;
+using MergerLogic.Utils;
 
 namespace MergerLogic.DataTypes
 {
     public class TMS : HttpDataSource
     {
-        public TMS(DataType type, string path, int batchSize, Extent extent, int maxZoom, int minZoom = 0, bool isOneXOne = false,
+        public TMS(IUtilsFactory utilsFactory, IOneXOneConvetor oneXOneConvetor,
+            DataType type, string path, int batchSize, Extent extent, int maxZoom, int minZoom = 0, bool isOneXOne = false,
```
Comment (shimoncohen) [inline MergerLogic/DataTypes/TMS.cs:9]: Same here (and all HTTP classes), no need to pass DataType.

File: MergerLogic/ImageProccessing/ITileScaler.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+using MergerLogic.Batching;
+using MergerLogic.DataTypes;
+
+namespace MergerLogic.ImageProccessing
+{
+    public interface ITileScaler
+    {
+        void Upscale(MagickImage baseImage, Tile baseTile, Coord targetCoords);
+    }
+}
```
Comment (shimoncohen) [inline MergerLogic/ImageProccessing/ITileScaler.cs:11]: Add line ;)

File: MergerLogic/Utils/IConfigurationManager.cs [hunk 6]
```diff
@@ -0,0 +1,7 @@
+﻿namespace MergerLogic.Utils
+{
+    public interface IConfigurationManager
+    {
+        string GetConfiguration(params string[] configPath);
+    }
+}
```
Comment (shimoncohen) [inline MergerLogic/Utils/IConfigurationManager.cs:7]: Add line ;)

File: MergerLogic/Utils/IOneXOneConvetor.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        Coord ToTwoXOne(Coord oneXOneCoords);
+        Coord ToTwoXOne(int z, int x, int y);
+        Tile ToTwoXOne(Tile tile);
+        Coord? TryFromTwoXOne(Coord cords);
+        Coord? TryFromTwoXOne(int z, int x, int y);
+        Tile? TryFromTwoXOne(Tile tile);
+        Coord? TryToTwoXOne(int z, int x, int y);
+        Tile? TryToTwoXOne(Tile tile);
+    }
+}
```
Comment (shimoncohen) [inline MergerLogic/Utils/IOneXOneConvetor.cs:20]: Add line ;)

File: MergerLogic/Extentions/ServiceCollectionExtentions.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
+﻿
```
Comment (shimoncohen) [inline MergerLogic/Extentions/ServiceCollectionExtentions.cs:1]: Remove empty lines.

File: MergerLogic/Utils/S3Utils.cs [hunk 9]
```diff
@@ -4,15 +4,18 @@
...
 namespace MergerLogic.Utils
 {
-    public class S3Utils : DataUtils
+    public class S3Utils : DataUtils, IS3Utils
     {
-        private AmazonS3Client client;
         private string bucket;
 
-        public S3Utils(AmazonS3Client client, string bucket, string path) : base(path)
+        private IAmazonS3 client;
```
Comment (shimoncohen) [inline MergerLogic/Utils/S3Utils.cs:11]: `_client`

**Review by shimoncohen — CHANGES_REQUESTED** [review]

File: MergerLogic/DataTypes/Fs.cs (hunk 1 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:12]: fixed

File: MergerLogic/DataTypes/Fs.cs (hunk 2 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:11]: fixed

File: MergerLogic/DataTypes/Fs.cs (hunk 3 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:10]: fixed

File: MergerLogic/DataTypes/Fs.cs [hunk 10]
```diff
@@ -3,17 +3,21 @@
...
         private bool done;
         private int completedTiles;
 
-        public FS(DataType type, string path, int batchSize, bool isOneXOne = false, bool isBase = false, GridOrigin origin = GridOrigin.LOWER_LEFT)
-            : base(type, path, batchSize, new FileUtils(path), isOneXOne, origin)
+        private IPathUtils _pathUtils;
+
+        public FS(IPathUtils pathUtils, IUtilsFactory utilsFactory, IOneXOneConvetor oneXOneConvetor,
+            DataType type, string path, int batchSize, bool isOneXOne = false, bool isBase = false, GridOrigin origin = GridOrigin.LOWER_LEFT)
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/Fs.cs:17]: done

File: MergerLogic/DataTypes/TMS.cs (hunk 4 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/DataTypes/TMS.cs:9]: fixed

File: MergerLogic/ImageProccessing/ITileScaler.cs (hunk 5 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/ImageProccessing/ITileScaler.cs:11]: fixed

File: MergerLogic/Utils/IConfigurationManager.cs (hunk 6 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IConfigurationManager.cs:7]: fixed

File: MergerLogic/Utils/IUtilsFactory.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+﻿namespace MergerLogic.Utils
+{
+    public interface IUtilsFactory
+    {
+        T GetDataUtils<T>(string path) where T : IDataUtils;
+        IFileUtils GetFileUtiles(string path);
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IUtilsFactory.cs:6]: fixed

File: MergerLogic/Utils/ITimeUtils.cs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+﻿
+namespace MergerLogic.Utils
+{
+    public interface ITimeUtils
+    {
+        void PrintElapsedTime(string prompt, TimeSpan ts);
+    }
+}
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/ITimeUtils.cs:8]: fixed

File: MergerLogic/Utils/IPathPatternUtils.cs [hunk 13]
```diff
@@ -0,0 +1,10 @@
+﻿using MergerLogic.DataTypes;
+
+namespace MergerLogic.Utils
+{
+    public interface IPathPatternUtils
+    {
+        string renderUrlTemplate(Coord coords);
+        string renderUrlTemplate(int x, int y, int z);
+    }
+}
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IPathPatternUtils.cs:10]: fixed

File: MergerLogic/Utils/IPathUtils.cs [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public interface IPathUtils
+    {
+        Coord FromPath(string path, bool isS3 = false);
+        string GetTilePath(string basePath, int z, int x, int y, bool isS3 = false);
+        string GetTilePath(string basePath, Tile tile);
+        string GetTilePathTMS(string basePath, int z, int x, int y, bool isS3 = false);
+        string RemoveTrailingSlash(string path, bool isS3 = false);
+    }
+}
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IPathUtils.cs:14]: fixed

File: MergerLogic/Utils/IOneXOneConvetor.cs (hunk 7 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IOneXOneConvetor.cs:20]: fixed

File: MergerLogic/Utils/IDataUtils.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace MergerLogic.Utils
+{
+    public interface IDataUtils
+    {
+        Tile GetTile(Coord coord);
+        Tile GetTile(int z, int x, int y);
+        bool TileExists(int z, int x, int y);
+    }
+}
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IDataUtils.cs:12]: fixed

File: MergerLogic/Utils/IPathPatternUtils.cs [hunk 16]
```diff
@@ -0,0 +1,10 @@
+﻿using MergerLogic.DataTypes;
+
+namespace MergerLogic.Utils
+{
+    public interface IPathPatternUtils
+    {
+        string renderUrlTemplate(Coord coords);
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IPathPatternUtils.cs:7]: fixed

File: MergerLogic/Utils/IPathPatternUtils.cs [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+        string renderUrlTemplate(int x, int y, int z);
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/IPathPatternUtils.cs:8]: fixed

File: MergerService/src/IRun.cs [hunk 18]
```diff
@@ -0,0 +1,7 @@
+﻿namespace MergerService.Src
+{
+    public interface IRun
+    {
+        void Start();
+    }
+}
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerService/src/IRun.cs:7]: fixed

File: MergerLogic/Extentions/ServiceCollectionExtentions.cs (hunk 8 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Extentions/ServiceCollectionExtentions.cs:1]: fixed

File: MergerLogic/Utils/GpkgUtils.cs [hunk 19]
```diff
@@ -6,20 +6,23 @@
 
 namespace MergerLogic.Utils
 {
-    public class GpkgUtils : DataUtils
+    public class GpkgUtils : DataUtils, IGpkgUtils
     {
         private string tileCache;
```
[the remark this answers is not in the document]
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/GpkgUtils.cs:11]: fixed

File: MergerLogic/Utils/S3Utils.cs (hunk 9 above)
Comment (cl-shaharshavit-rnd) [reply MergerLogic/Utils/S3Utils.cs:11]: fixed

**Review by asafMasa — APPROVED** [review]

**Review by shimoncohen — APPROVED** [review]

**Pull request merged**